ikarus/finiteelements/mechanics/materials/materialhelpers.hh

```diff
@@ -0,0 +1,185 @@
+  std::ranges::set_difference(std::ranges::iota_view(size_t(0), size_t(6)), voigtFixedIndices, res.begin());
+  std::ranges::sort(res);
+  return res;
+}
+
+/**
+ * \brief Helper function to create an array of fixed Voigt indices.
+ * \tparam size The size of the fixed pairs array.
+ * \param fixed An array of MatrixIndexPair representing fixed indices.
+ * \return std::array<size_t, size> The array of fixed Voigt indices.
+ */
+template <size_t size>
+consteval auto createFixedVoigtIndices(const std::array<MatrixIndexPair, size>& fixed) {
+  std::array<size_t, size> fixedIndices;
+  std::ranges::transform(fixed, fixedIndices.begin(), [](auto pair) { return toVoigt(pair.row, pair.col); });
+  std::ranges::sort(fixedIndices);
+  return fixedIndices;
+}
+
+/**
+ * \brief Helper function to count the number of diagonal indices in the fixed pairs array.
+ * \tparam size The size of the fixed pairs array.
+ * \param fixed An array of MatrixIndexPair representing fixed indices.
+ * \return constexpr size_t The number of diagonal indices.
+ */
+template <size_t size>
+constexpr size_t countDiagonalIndices(const std::array<MatrixIndexPair, size>& fixed) {
+  size_t count = 0;
+  for (auto v : fixed) {
+    if (v.col == v.row)
+      ++count;
+  }
+  return count;
+}
+
+/**
+ * \brief Converts the input strain matrix to the appropriate form for stress reduction.
+ * \tparam Derived The derived type of the input matrix.
+ * \param E The input strain matrix.
+ * \return decltype(auto) The converted strain matrix.
+ */
+template <typename Derived>
+decltype(auto) maybeFromVoigt(const Eigen::MatrixBase<Derived>& E) {
+  if constexpr (Concepts::EigenVector<Derived>) { // receiving vector means Voigt notation
+    return fromVoigt(E.derived(), true);
+  } else
+    return E.derived();
+}
+
+template <typename ScalarType>
+void checkPositiveDet(ScalarType det) {
+  if (Dune::FloatCmp::le(static_cast<double>(det), 0.0, 1e-10))
+    DUNE_THROW(Dune::InvalidStateException,
+               "Determinant of right Cauchy Green tensor C must be greater than zero. detC = " +
+                   std::to_string(static_cast<double>(det)));
+}
+
+/**
+ * \brief Computes the principal stretches of the input strain matrix C.
+ *
+ * \tparam ScalarType The ScalarType
+ * \tparam Derived The derived type of the input matrix
+ * \param C the input strain matrix
+ * \param options should be either `Eigen::ComputeEigenvectors` or `Eigen::EigenvaluesOnly`
+ * \return auto pair of principalstretches and corresponding eigenvectors (if `Eigen::EigenvaluesOnly` the
+ * eigenvectors is a zero matrix )
+ */
+template <typename ScalarType, typename Derived>
+auto principalStretches(const Eigen::MatrixBase<Derived>& C, int options = Eigen::ComputeEigenvectors) {
+  Eigen::SelfAdjointEigenSolver<Derived> eigensolver{};
+
+  // For AD we use the direct method which uses a closed form calculation, but has less accuracy
+  if constexpr (Concepts::AutodiffScalar<ScalarType>)
```

Comment:
It might be working if you dont use computeDirect? Because this might be given better results for the deviatoric degeneracy issue.